### ISSUE TITLE
Firebase fetches a new token on each request

### ISSUE THREAD

**Issue by peterp** [redwoodjs/redwood#2502]
https://github.com/redwoodjs/redwood/blob/3557f707889998ccdc1c3542048618b2ef8ba6ac/packages/auth/src/authClients/firebase.ts#L56

You would:
1. Have a place to store the token's payload ( with expiration time )
2. Compare the current time against the expiration time whenever getToken is called
3. If it has not expired, return the old token, if it has expired, call the part the retrieves the token again.

![image](https://user-images.githubusercontent.com/44849/118547471-a5689900-b759-11eb-89ea-e6cd038d1bf2.png)

**Comment by callingmedic911** [comment]
@peterp I was trying to replicate issue - that there's new token generated if we navigate/refresh page. Right?
But I'm not able to replicate the issue, so far I did: 
- Cloned [redwood-tutorial](https://github.com/redwoodjs/redwood-tutorial)
- Setup [firebase auth](https://redwoodjs.com/docs/authentication#firebase-2) with provider as Google.

I see only single token call (after login/expiring after a while) even after navigating/refreshing multiple times.
![Screenshot 2021-05-27 at 11 00 33 PM](https://user-images.githubusercontent.com/2629902/119870761-597ad880-bf3f-11eb-901a-bfe7b7e03928.png)


I looked up Firebase SDK docs to understand it better, and there I see that [`getIdToken`](https://firebase.google.com/docs/reference/js/firebase.User#getidtoken) is itself supposed to return current token if not expired. Also, I see the Indexed Db auth related data is stored as well.
![Screenshot 2021-05-27 at 11 02 51 PM](https://user-images.githubusercontent.com/2629902/119871164-cf7f3f80-bf3f-11eb-9822-1f69ca4a820d.png)

Looks like I'm missing something here. Can you please help?

**Comment by thedavidprice** [comment]
^^ @dthyresson do you have any Firebase knowledge about this?

**Comment by dthyresson** [comment]
I'd have to test on Auth playground - I don't use Firebase auth in my apps.

When I did try I didn't see that behavior. Might it be specific to the UserVitals implementation?

It would be nice to replicate this issue on https://redwood-playground-auth.netlify.app/firebase.

**Comment by viperfx** [comment]
Hmm - I will review the code we have. Not sure what we are doing anything out of the ordinary that would make this happen. I wonder if it's specific to the react-select and the way we are doing the query for auto-complete.

I was looking into doing something similar to #2555 to resolve this - but if it doesn't happen for others then I will investigate on my side first before doing anything. I think Danny also mentioned he is not seeing it.

This is the function we use with react-select to do the querying 

```
  const search = useCallback(
    AwesomeDebouncePromise(async (query) => {
      if (query === '') {
        const contactData = await recentContacts.refetch()
        return contactData?.data?.recentContacts
      }
      const { data } = await searchQuery.refetch({ query: query })
      const contacts = data?.searchFeedbackUser.map((f) => {
        return { email: f.email, name: f.name, contactId: f.id, role: f.role }
      })
      return contacts
    }, 150),
    []
  )
```

**Comment by doesnotexist** [comment]
I would have thought that firebase's setPersistence (https://firebase.google.com/docs/auth/web/auth-state-persistence#web-version-9_1) which defaults to local storage would govern whether getIdToken makes a request each time or only when needing to refresh the token.  Anyhow, I'll see if I can reproduce this behavior.

**Comment by thedavidprice** [comment]
@doesnotexist curious if you have any updates here based on your work in #3347? Thanks!

**Comment by doesnotexist** [comment]
I was not able to reproduce this behavior, but I've only been testing it against the v9 implementation in the pull request, but that probably isn't related anyhow.  Since regardless of v8 or v9, calling getIdToken is correct implementation for the auth client's getToken() method and should not make any requests if the locally persisted token has not or is not going to expire soon (within 5 minutes according to the v9 docs)
v8: https://firebase.google.com/docs/reference/js/v8/firebase.User#getidtoken
v9: https://firebase.google.com/docs/reference/js/auth.user.md#usergetidtoken

I am able to provoke something like the behavior by changing the implementation to call getIdToken(true), passing the optional argument forceRefresh=true and that results in multiple network requests on page loads  https://securetoken.googleapis.com/v1/token?key=AI....

In the unmodified implementation, I usually only see a single request to a google endpoint on initial page load when the AuthProvider componentDidMount() lifecycle method is called and the firebase auth object during initialization, also when this happens the endpoint is slightly different as it hits /accoutns:lookup and not /token
 https://identitytoolkit.googleapis.com/v1/accounts:lookup?key=AI...
As a result it appears to set the local persisted token and prevents the subsequent calls to getIdToken() from making those /token?key=AI...  requests.

All of this behavior is dictated by the firebase client's internals and I have trouble coming up with anything that is happening on the redwood side that would cause this unexpected behavior.   In the #3347 @viperfx mentioned that they are using CustomTokens with firebase.  Which I assume is done in this fashion
https://firebase.google.com/docs/auth/admin/create-custom-tokens#create_custom_tokens_using_the_firebase_admin_sdk
using the firebase admin sdk on their backend.   Perhaps there is a bug regarding token expiration times when using customTokens and if so that would be a bug in firebase and not in redwood.   Maybe they can check to see if the issue is related to token expiration by adding some logging like so:
```ts
     getToken: async () => {
      const token = auth.currentUser ? auth.currentUser.getIdToken() : null
      if (token && auth.currentUser) {
        const tokenResult = await auth.currentUser.getIdTokenResult()
        console.log(`expiration time = ${tokenResult.exp}`)
        console.log(`expiration time = ${tokenResult.expirationTime}`)
      }
      return token
    },
```

**Comment by viperfx** [comment]
So yes, I don't believe this happening anymore. It was most noticeable in our auto complete fields but the noticed originally in this PR is no longer. 

Nevertheless, will take your debug output above and make sure this is no longer an issue on my side.

**Comment by thedavidprice** [comment]
Thank you @doesnotexist

@viperfx please do let us know if you get a chance to confirm.

**Comment by viperfx** [comment]
Yep - so can confirm that this is fixed on my end in my current rw app with v0.37

**Comment by thedavidprice** [comment]
Closing now! 

v0.37.0-rc will be out later today. We can re-open if/as needed.

Thanks all!
